### PR TITLE
fabio 1.0.8 (new formula)

### PR DESCRIPTION
**Pull request by svdberg**
Add formula for Fabio (https://github.com/eBay/fabio), a zero-conf load balancing HTTP(S) router. Written in Go (golang), using consul to store url -> ip:port mappings.

Added an optional, but recommended Consul dependency.

### CODE REVIEW

File: Library/Formula/fabio.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    ENV["GOPATH"] = "#{buildpath}/_third_party:#{buildpath}"
+
+    Language::Go.stage_deps resources, buildpath/"src"
+
+    system "go", "build", "-o", "fabio"
+    bin.install "fabio"
+  end
+
+  test do
+    output = shell_output("#{bin}/fabio -v")
```
Comment (mikemcquaid) [inline Library/Formula/fabio.rb]: Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

Comment (svdberg) [inline Library/Formula/fabio.rb]: This is a network daemon, that depends on Consul to be able to start up. If I would do a "functional" test, I'd have to run Consul or make sure it is running. Can I start consul from within the sandbox?

Comment (mikemcquaid) [inline Library/Formula/fabio.rb]: Yeh, you can use `fork do` to start a new process. If you `grep` for that you should find some other examples of similar usage.

**Comment by svdberg** [comment]
hmm Travis seems to get stuck on this one..

**Comment by svdberg** [comment]
@mikemcquaid Any idea why the default wont turn green?

**Comment by mikemcquaid** [comment]
@svdberg It'll take a while, our CI infrastructure is kinda overloaded.

File: Library/Formula/fabio.rb (hunk 1 above)
Comment (svdberg) [inline Library/Formula/fabio.rb]: Added new test, see updated pull request.

File: Library/Formula/fabio.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    if Fabio.port_open?(LOCALHOST_IP, CONSUL_DEFAULT_PORT) && !Fabio.port_open?(LOCALHOST_IP, FABIO_DEFAULT_PORT)
+      # Consul is there..
+      fork do
+        exec "#{bin}/fabio &>fabio-start.out&"
+      end
+      sleep 10
+      assert_equal Fabio.port_open?(LOCALHOST_IP, FABIO_DEFAULT_PORT), true
+      exec "killall fabio" # fabio forks off from the fork...
+    else
+      # consul not running or fabio already running (or something else on that port), fallback to version test
```
Comment (mikemcquaid) [inline Library/Formula/fabio.rb]: I think it's fine for the test to just fail in this case.

File: Library/Formula/fabio.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    ENV["GOPATH"] = "#{buildpath}/_third_party:#{buildpath}"
+
+    Language::Go.stage_deps resources, buildpath/"src"
+
+    system "go", "build", "-o", "fabio"
+    bin.install "fabio"
+  end
+
+  test do
+    # check if Consul is reachable
```
Comment (mikemcquaid) [inline Library/Formula/fabio.rb]: Can skip all the comments, they are pretty self-explanatory.

**Comment by mikemcquaid** [comment]
Nice test :+1: A few comments and this is ready to :ship:

**Comment by svdberg** [comment]
Hmm.. Now test fail because Consul is not running. I can't start consul because of sandbox and the *optional* dependency on Consul. I can re-add the version check test as fallback, that would make the test pass. What is your opinion @mikemcquaid ?

File: Library/Formula/fabio.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      sleep 10
+      assert_equal true, Fabio.port_open?(LOCALHOST_IP, FABIO_DEFAULT_PORT)
+      exec "killall fabio" # fabio forks off from the fork...
+    else
+      puts "Fabio already running or Consul not available or starting fabio failed."
+      false
+    end
+  end
+
+  def self.port_open?(ip, port, seconds = 1)
```
Comment (mikemcquaid) [inline Library/Formula/fabio.rb]: It's be good to make this a non-class method and put it inside the `test do` block.

**Comment by mikemcquaid** [comment]
@svdberg It sounds like it's a mandatory dependency in that case? Regardless, the test is for the default configuration so it'd be OK to have Consol run in the test.

**Comment by svdberg** [comment]
All done, all green.

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request closed without merging**